### ISSUE TITLE
numpy/scipy adjacency matrix to Graph

### ISSUE THREAD

**Issue by Jessime** [igraph/python-igraph#168]
I have a large (~300M edges) adjacency matrix (either as a `np.array` or a `scipy.sparse.csr_matrix`, depending on the density of edges) that I'd like to call communities on. Is there a convenient way to convert either of these objects to a `Graph` object?

Thanks!

**Comment by ntamas** [comment]
Call `.nonzero()` on your matrix to find out the indices of the nonzero elements, and then use the standard `Graph()` constructor:

```python
sources, targets = your_matrix.nonzero()
edgelist = zip(sources.tolist(), targets.tolist())
g = Graph(edgelist)
```

**Comment by Jessime** [comment]
I tried running:

```python
graph = igraph.Graph(zip(sim_sparse.row.tolist(), sim_sparse.col.tolist()), 
                     edge_attrs={'weight': sim_sparse.data.tolist()})
```

and still have a few questions:

1. Most importantly, this gives me: `TypeError: an integer is required (got type zip)`. Running `igraph.__version__` gives me '0.7.1', if that matters. 
2. The 300M edge count is only counting nonzero elements, so I'd love to avoid going to a pure python list if at all possible, since my memory usage explodes. 
3. Am I providing edge weights properly?

**Comment by ntamas** [comment]
1. This is probably because you are on Python 3.x where `zip` returns an iterator and not a list. `igraph.Graph` expects a list of vertex pairs as its first argument (or an integer, in which case it is interpreted as the number of nodes and the edge list becomes the second argument). So, convert the iterator into a list with `list(zip(...))`.

2. Unfortunately that's not possible; deep down is a C library behind igraph, and the graph constructor in that library expects an edge list, not a Python iterator. Somewhere, at some point, one has to bite the bullet and convert the sparse matrix into an edge list. You can work around it with temporary files on the disk. First, you need to convert the matrix composed of `sim_sparse.row` and `sim_sparse.col` (both as column vectors) into a file named `edges.txt` (use `np.savetxt()` for this), and the weights into another file named `weights.txt` (again, you can use `np.savetxt()`). Then you can delete the matrix (`del sim_sparse`), construct the graph with `g = Graph.Read_Edgelist("edges.txt")`, and attach the weights by loading them back from `weights.txt` into a list and then calling `g.es["weight"] = weights`. If your graph is undirected, you should probably mask the lower triangle of the sparse matrix before saving it so you don't save each edge twice.

3. Yes, you are.

**Comment by xinzhige** [comment]
The method only considers the nodes with edges (in the edgelist). If there are nodes without any edges, the graph created would not contain these nodes. So the size of the graph would be smaller, which is not what we want.

**Comment by ntamas** [comment]
I assume that you are referring to [my comment on Mar 20](https://github.com/igraph/python-igraph/issues/168#issuecomment-374520102). If that is the case, the following variant preserves the node count even if the last few nodes in the matrix are isolates:

```python
vcount = max(your_matrix.shape)
sources, targets = your_matrix.nonzero()
edgelist = zip(sources.tolist(), targets.tolist())
g = Graph(vcount, edgelist)
```

**Comment by ntamas** [comment]
Yes, if you are on Python 3.x, you need to wrap it in `list()`. On Python 2.x this is not a problem because `zip()` returns a list there.

**Comment by mollyhjorthjensen** [comment]
When I try and do this I get the error: "Graph should not be instantiated directly", why is that?

**Comment by gnool** [comment]
Is there anyway to add edges without converting numpy array to Python list via tolist() ? This is a really expensive operation for large graph. Also, why is there a need to convert to Python list when the backend is anyway C/C++? Isn't a numpy array already in a format that the backend can use? In other words, is there any way we can do add_edges(edges), where edges is a Nx2 numpy array.

**Comment by ntamas** [comment]
> Is there anyway to add edges without converting numpy array to Python list via tolist() ?
No, there isn't.

> Isn't a numpy array already in a format that the backend can use?
It depends. _If_ the NumPy array contains double-precision floats, _and_ its size is N rows by 2 columns, _and_ the items are stored in row-major order, _then_ the format is in theory usable directly by igraph. However, in order to do that, we would need to obtain a pointer to the internal memory using [`PyArray_DATA()`](https://www.numpy.org/devdocs/reference/c-api.array.html#c.PyArray_DATA) from the C layer, which means that we would need the NumPy headers at compilation time -- effectively making NumPy a dependency of `python-igraph` even if one has no intention to use NumPy at all.

**Comment by gnool** [comment]
@ntamas Thanks for your explanation. I may have a wrong understanding of how python-igraph was intended for those who look for speed (in both construction of graph and subsequent uses of it). python-igraph was indeed fast when I perform shortest path search on my dataset, but the construction of graph was a bit too slow for my application. In the end I found scipy's csgraph was actually applicable for my application, there both graph construction and path finding were as fast as it could get, and I did not have to worry about converting edges in numpy array to Python list first.

**Comment by ntamas** [comment]
> I may have a wrong understanding of how python-igraph was intended for those who look for speed
Well, mainly that's the goal, but most of the speed of `python-igraph` actually comes from the underlying C library, and for most of the time I was negligent about the performance implications of crossing the Python-C boundary (because in 99% of the cases the amount of data moved through that boundary is small - maybe only the graph construction is the only case where this is not true).

In the meanwhile, I was looking into what `numpy` has to offer in its Python API for getting access to the underlying arrays, and I have found the `numpy.ctypeslib` module (see source [here](https://github.com/numpy/numpy/blob/master/numpy/ctypeslib.py) ). If I understand correctly, this should allow me to get hold of a `ctypes` array pointer to the underlying data. The good thing about this is that `ctypes` is a standard Python library, so I can add some glue code on the C layer of `python-igraph` that allows the construction of a graph from an edge list provided as a `ctypes` pointer, and handle the conversion from NumPy arrays to a `ctypes` pointer in the Python layer (where it is easy to import `numpy` directly without adding a compile-time dependency on `numpy`).

Judging from you response, it seems like you have already found an alternative solution that you are happy with, and considering that I don't really have much free time to dedicate to new feature requests in `python-igraph`, I won't start working on the implementation right now, but I'll leave it for the future when I can find a few spare hours to investigate this further.

**Comment by Jessime** [comment]
For what it's worth, I would be as excited about this change for the benefits in memory usage as speed. I'm often working with an adjacency matrix and/or graph that's just large enough to fit into my laptop's memory when it's stored as a numpy array. Depending on the specifics, conversion to a list is a non-starter since the memory usage is going to make my laptop grind to a halt when it runs out of swap. Usually the work-around is moving all my data to a remote machine, which is a hassle.

But this seems like a fairly serious undertaking, so I definitely understand if it's not a quick fix. Just wanted to share another use case for numpy arrays!

**Comment by ntamas** [comment]
A first version of this is now committed in fdbf1182; this version supports creating graphs from NumPy arrays and matrices. The C part of the glue code actually supports creating graphs from Python `memoryview` objects and checks a few basic requirements about the `memoryview` object (it must be 2D with 2 columns, its item size must match the native item size of igraph vectors, and it must be contiguous in memory). Then there is some additional code in the Python layer that recognizes NumPy arrays and matrices dynamically at runtime, and ensures that they are laid out in a way that is okay for the C core of igraph (and makes a copy if this is not the case).

**Comment by Jessime** [comment]
Nice! This is a big improvement to igraph's overall usability in Python. Thanks!

**Comment by liliblu** [comment]
I'm very excited about this feature too! I don't see the #fdbf118 code change available in a release, is that in the works anywhere? or maybe I'm installing from the wrong place? I'm using conda. I would like to use it as a dependency, so it would be great if there were a version with this change.

**Comment by ntamas** [comment]
This is not in the released version yet, but `python-igraph` 0.8 will be released in the next few weeks, so stay tuned :) Not sure when the new version will be included in conda, though.

In the meanwhile, if you have a working C compiler on your machine and you don't mind compiling from source, you can install the development version from git as follows:

```
pip install -e git+https://github.com/igraph/python-igraph#egg=python-igraph
```

**Comment by liliblu** [comment]
This is really exciting! thank you!
